testenv/utils/ssh.go

```diff
@@ -0,0 +1,291 @@
+	LocalPort     int
+	RemoteAddress string
+	RemotePort    int
+
+	Callback SSHForwardCallback
+}
+
+func (i *SSHForwardInput) HostAddressPort() string {
+	return fmt.Sprintf("%s:%d", i.Host, i.Port)
+}
+
+func (i *SSHForwardInput) LocalAddressPort() string {
+	return fmt.Sprintf("localhost:%d", i.LocalPort)
+}
+
+func (i *SSHForwardInput) RemoteAddressPort() string {
+	if i.RemoteAddress == "" {
+		i.RemoteAddress = "localhost"
+	}
+
+	return fmt.Sprintf("%s:%d", i.RemoteAddress, i.RemotePort)
+}
+
+type SSHPortForward struct {
+	input        *SSHForwardInput
+	clientConfig ssh.ClientConfig
+
+	stop atomic.Bool
+}
+
+func (f *SSHPortForward) Start(ctx context.Context) error {
+	// Dial the remote server.
+	client, err := ssh.Dial("tcp", f.input.HostAddressPort(), &f.clientConfig)
+	if err != nil {
+		return fmt.Errorf("failed to dial:%w", err)
+	}
+
+	// Listen on local port.
+	listener, err := net.Listen("tcp", f.input.LocalAddressPort())
+	if err != nil {
+		return fmt.Errorf("failed to listen: %w", err)
+	}
+
+	// TODO: wait until the BannerCallBack is invoked
+
+	go func() {
+		var callbackErr error
+		for callbackErr == nil && !f.stop.Load() {
+			defer client.Close()
+			defer listener.Close()
+
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			// Accept local connection.
+			local, err := listener.Accept()
+			if err != nil {
+				callbackErr = f.input.Callback(ctx, fmt.Errorf("local SSH tunnel error: %w", err))
+			}
+			defer local.Close()
+
+			// Dial remote server.
+			remote, err := client.Dial("tcp", f.input.RemoteAddressPort())
+			if err != nil {
+				callbackErr = f.input.Callback(ctx, fmt.Errorf("remote SSH tunnel error: %w", err))
+			}
+			defer remote.Close()
+
+			wg := sync.WaitGroup{}
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+
+				_, err := io.Copy(local, remote)
+				if err != nil {
+					callbackErr = f.input.Callback(ctx, fmt.Errorf("failed to copy data from remote to local: %w", err))
+				}
+			}()
+
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+
+				_, err := io.Copy(remote, local)
+				if err != nil {
+					callbackErr = f.input.Callback(ctx, fmt.Errorf("failed to copy data from remote to local: %w", err))
+				}
+			}()
+
+			wg.Wait()
+		}
+	}()
+
+	return nil
+}
+
+func (f *SSHPortForward) Stop() {
+	f.stop.Store(true)
+}
+
+func NewSSHPortForward(input *SSHForwardInput) (*SSHPortForward, error) {
+	if input.Callback == nil {
+		input.Callback = func(ctx context.Context, err error) error {
+			logger := GetLoggerFromContextOrDiscard(ctx)
+			logger.Error(err)
+
+			return err
+		}
+	}
+
+	// Create Signer from private key.
+	signer, err := ssh.ParsePrivateKey(input.PrivateKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse private key: %w", err)
+	}
+
+	// Create SSH client config.
+	config := ssh.ClientConfig{
+		User: input.User,
+		Auth: []ssh.AuthMethod{
+			ssh.PublicKeys(signer),
+		},
+		HostKeyCallback:   ssh.InsecureIgnoreHostKey(), //nolint:gosec
```

Comment:
Did you try verifying host key?
```suggestion
		HostKeyCallback:   ssh.FixedHostKey(signer.PublicKey()),
```

Reply:
this change fails with `failed to run after setup: failed to wait for the SSH port to become ready: failed to dial:ssh: handshake failed: ssh: host key mismatch`